### PR TITLE
Store etag So Downloads Use Cache

### PR DESCRIPTION
**Pull request by kunalgoyal9**
Storing Etag received from `fetch` API parallel to mtime file and passing this header to `If-None-Match`. 
Fixing https://github.com/teaxyz/cli/issues/85

### CODE REVIEW

File: src/hooks/useDownload.ts
Comment (mxcl) [inline src/hooks/useDownload.ts:48]: This means we'll only do these if BOTH etag and mtime are provided. ETAGs are much rarer than mtimes.

Comment (kunalgoyal9) [reply src/hooks/useDownload.ts:48]: Can we do this if any of mtime or etag condition met?

Comment (mxcl) [reply src/hooks/useDownload.ts:48]: yes, if any of them are met. I recommend reading up on how HTTP caching works, I do not have time to teach you. Thanks.

**Comment by mxcl** [comment]
Needs a test

File: src/hooks/useDownload.ts
Comment (mxcl) [inline src/hooks/useDownload.ts:51]: https://stackoverflow.com/a/10671877

Don’t do both. Please research the task at hand before attempting it.

**Review by mxcl — CHANGES_REQUESTED** [review]
Not following best practices.

File: tests/unit/cache.test.ts
Comment (mxcl) [inline tests/unit/cache.test.ts:7]: This is a 30MB file, pick something less than 100B pls.

Also, this is fragile. We should not rely on the Internet for testing. Really you should spin up a local web server using a tea-script and test against that.

File: src/hooks/useDownload.ts
Comment (kunalgoyal9) [reply src/hooks/useDownload.ts:51]: Sure... I will. Thanks for the clarification

File: tests/unit/cache.test.ts
Comment (mxcl) [reply tests/unit/cache.test.ts:7]: the work required for that test infrastructure is pretty high tho. However tests failing due to flakey network conditions is the worst.

However, it will pay off. We can use it to test just about everything.

Comment (jhheider) [reply tests/unit/cache.test.ts:7]: It might be as simple as `tea -X npx http-server test.fixtures/ -p 2051`

Caveats:
- 2051 = the ordinals for T(20) E(5) A(1)
- 2051 is `epnsdp`, which a quick google search might be an archaic dns thing, and used by some commercial backup solutions in their products (the port, not the service)
- the `tea -X npx` command will *also* flake if there's internet issues, but it allows a single install call before any number of http tests

File: src/hooks/useDownload.ts
Comment (kunalgoyal9) [reply src/hooks/useDownload.ts]: Fixed

File: README.md
Comment (mxcl) [inline README.md:805]: why?

File: src/hooks/useDownload.ts
Comment (mxcl) [inline src/hooks/useDownload.ts:51]: our code format is space before the parameter after the `if`

File: tests/unit/cache.test.ts
Comment (mxcl) [inline tests/unit/cache.test.ts:8]: the `dst` is wrong. Just remove it.

File: tests/unit/cache.test.ts
Comment (mxcl) [inline tests/unit/cache.test.ts:11]: so we're testing an implementation detail of how useDownload() works?

Basically this test verifies useDownload() but it doesn't really test that etags work.

**Review by mxcl — CHANGES_REQUESTED** [review]
Approved after the mods

File: src/hooks/useDownload.ts
Comment (jhheider) [inline src/hooks/useDownload.ts:91]: Supremely minor: I'd use consistent spacing on object notation on parallel lines.

I know our codebase as it stands would currently fail, but do you still like to "see how [you] feel" w.r.t. object notation spacing in TS, @mxcl, or should we adopt rigid guidelines? (I think we could also make this happen via eslint/prettier/etc rules and auto-application. This might prevent me from introducing semi-colons (though I don't believe I have done this in some time), and we might be able to catch imports which don't use the import map, though that might require some custom logic.)

File: README.md
Comment (kunalgoyal9) [reply README.md:805]: Because the test I have created need write access to `~/.tea/tea.xyz/var/www/https/dist.tea.xyz/` directory for writing mtime and etag files

File: tests/unit/cache.test.ts
Comment (kunalgoyal9) [reply tests/unit/cache.test.ts:8]: if there is no `dst` specified it will download it to `tea` directory i.e. `~/.tea/tea.xyz/var/www/https/dist.tea.xyz`

File: src/hooks/useDownload.ts
Comment (kunalgoyal9) [reply src/hooks/useDownload.ts]: Fixed

File: tests/unit/cache.test.ts
Comment (mxcl) [reply tests/unit/cache.test.ts:8]: well I guess you're right in that you shouldn't do that, but you are currently downloading to a file called `test.tar.xz` which is flagrantly wrong. Anyone reading this code in the future will wonder what the fuck you were thinking.

File: README.md
Comment (mxcl) [reply README.md:805]: So set the new `--allow-write` to `--allow-write="$TMPDIR:$TEA_PREFIX/tea.xyz/var/www".

We don’t allow EVERYTHING just because one new thing needs more, you keep things as restrictive as possible.

Comment (mxcl) [reply README.md:805]: Frankly we should change TEA_PREFIX for the tests anyway.

File: tests/unit/cache.test.ts
Comment (mxcl) [inline tests/unit/cache.test.ts:8]: `versions.txt` not `version.txt`. This is sloppy.

File: README.md
Comment (jhheider) [reply README.md:805]: Comma, I think, not colon. You've spent too much time on the POSIX work.

Comment (kunalgoyal9) [reply README.md]: Fixed it, It is working on local but tests are getting failed. 

Seems like this is another issue with `$SRCROOT`. Tests that were working fine earlier are now failing :(

File: src/hooks/useDownload.ts
Comment (mxcl) [inline src/hooks/useDownload.ts:88]: Case sensitivity doesn't matter, but we capitalized it like the docs do on the previous line so consistency would have been `ETag` per the docs for ETags.

**Comment by mxcl** [comment]
In the end, this was good work, thanks @kunalgoyal9

**Pull request merged**
